package/MDAnalysis/analysis/dssp/dssp.py

```diff
@@ -0,0 +1,404 @@
+        atom coordinates for the secondary structure determination. The last
+        string contains multiple values for "O" to account for C-term residues.
+    hydrogen_name : str, default "H HN HT1 HT2 HT3"
+        This selection should only select a single hydrogen atom in each residue
+        (except proline), namely the one bound to the backbone nitrogen.
+
+    .. Note::
+       To work with different hydrogen-naming conventions by default, the default
+       selection is broad but if hydrogens are incorrectly selected (e.g., a
+       :exc:`ValueError` is raised) you must customize ``hydrogen_name`` for your
+       specific case.
+
+    Raises
+    ------
+    ValueError
+        if ``guess_hydrogens`` is True but some non-PRO hydrogens are missing.
+
+    Examples
+    --------
+
+    For example, you can assign secondary structure for a single PDB file:
+
+    >>> from MDAnalysis.analysis.dssp import DSSP
+    >>> from MDAnalysisTests.datafiles import PDB
+    >>> import MDAnalysis as mda
+    >>> u = mda.Universe(PDB)
+    >>> run = DSSP(u).run()
+    >>> print("".join(run.results.dssp[0][:20]))
+    --EEEEE-----HHHHHHHH
+
+    (Note that for displaying purposes we only print the first 20 residues
+    of frame 0 with ``run.results.dssp[0][:20]`` but one would typically look
+    at all residues ```run.results.dssp[0]``.)
+
+    Also, per-frame dssp assignment allows you to build average
+    secondary structure -- ``DSSP.results.dssp_ndarray`` holds
+    (n_frames, n_residues, 3) shape ndarray with one-hot encoding of
+    loop, helix and sheet, respectively:
+
+    >>> from MDAnalysis.analysis.dssp import translate, DSSP
+    >>> from MDAnalysisTests.datafiles import TPR, XTC
+    >>> u = mda.Universe(TPR, XTC)
+    >>> long_run = DSSP(u).run()
+    >>> mean_secondary_structure = translate(long_run.results.dssp_ndarray.mean(axis=0))
+    >>> print(''.join(mean_secondary_structure)[:20])
+    -EEEEEE------HHHHHHH
+
+    .. versionadded:: 2.8.0
+    """
+
+    def __init__(
+        self,
+        atoms: Union[Universe, AtomGroup],
+        guess_hydrogens: bool = True,
+        *,
+        heavyatom_names: tuple[str] = ("N", "CA", "C", "O O1 OT1"),
+        hydrogen_name: str = "H HN HT1 HT2 HT3",
+    ):
+        self._guess_hydrogens = guess_hydrogens
+
+        ag: AtomGroup = atoms.select_atoms("protein")
+        super().__init__(ag.universe.trajectory)
+
+        # define necessary selections
+        self._heavy_atoms: dict[str, "AtomGroup"] = {
+            t: ag.atoms[
+                np.isin(
+                    ag.names, t.split()
+                )  # need split() since `np.isin` takes an iterable as second argument
+                # and "N".split() -> ["N"]
+            ]
+            for t in heavyatom_names
+        }
+        self._hydrogens: list["AtomGroup"] = [
+            res.atoms.select_atoms(f"name {hydrogen_name}") for res in ag.residues
+        ]
```

Comment:
From the following code I see that there's the assumption that the **order of atoms** in `_heavy_atoms["CA"]` is the same as in `_hydrogens`.

However, that may not be true: `_hydrogens` is created from a `select_atoms()` which **always orders the atoms by index in increasing order** whereas ag's in `_heavy_atoms` are created by look up and slicing. **If the original input `ag` is NOT ORDERED then `_hydrogens` will be ORDERED whereas `_heavy_atoms` remain in the CUSTOM ORDER of the original ag**. At least that's what it looks like to me.

It may be safer to just sort the original `ag` first.

Reply:
>It may be safer to just sort the original ag first.

but there's line 274 that says:

`ag: AtomGroup = atoms.select_atoms("protein")`

and then all operations are on this `ag`. I even think that you suggested this, but I'm not entirely sure :)

Reply:
You're right: `ag` is guaranteed to be ordered.

Maybe add a comment somewhere that the code is guaranteed to work _because_ `ag` is ordered. Just in case someone later wants to make it more general (eg so that one doesn't have to rely on MDAnalysis's definition of "protein").

Reply:
See my comment on Dr. Richard Hipp's obsession with code comments https://discord.com/channels/807348386012987462/807712498249236520/1237819983833337917